libcni/api.go

```diff
@@ -193,24 +206,78 @@ func (c *CNIConfig) getResultCacheFilePath(netName string, rt *RuntimeConf) stri
 	return filepath.Join(c.getCacheDir(rt), "results", fmt.Sprintf("%s-%s-%s", netName, rt.ContainerID, rt.IfName))
 }
 
-func (c *CNIConfig) setCachedResult(result types.Result, netName string, rt *RuntimeConf) error {
+func (c *CNIConfig) setCachedResult(result types.Result, config []byte, netName string, rt *RuntimeConf) error {
+
+	cached := cachedInfo{
+		Config:     config,
+		Kind:       "cniCacheV1",
+		CNIVersion: "0.4.0"}
+	if len(rt.Args) > 0 {
+		cached.CniArgs = rt.Args
+	}
+	if len(rt.CapabilityArgs) > 0 {
+		cached.CapabilityArgs = rt.CapabilityArgs
+	}
+
+	// We need to get type.Result into cachedInfo as JSON map
+	// Marshal to []byte, then Unmarshal into cached.RawResults
 	data, err := json.Marshal(result)
 	if err != nil {
 		return err
 	}
+
+	err = json.Unmarshal(data, &cached.RawResults)
+	if err != nil {
+		return err
+	}
+
+	newBytes, err := json.Marshal(&cached)
+	if err != nil {
+		return err
+	}
+
 	fname := c.getResultCacheFilePath(netName, rt)
 	if err := os.MkdirAll(filepath.Dir(fname), 0700); err != nil {
 		return err
 	}
-	return ioutil.WriteFile(fname, data, 0600)
+
+	return ioutil.WriteFile(fname, newBytes, 0600)
 }
 
 func (c *CNIConfig) delCachedResult(netName string, rt *RuntimeConf) error {
 	fname := c.getResultCacheFilePath(netName, rt)
 	return os.Remove(fname)
 }
 
-func (c *CNIConfig) getCachedResult(netName, cniVersion string, rt *RuntimeConf) (types.Result, error) {
+func (c CNIConfig) getCachedConfig(netName string, rt *RuntimeConf) ([]byte, *RuntimeConf, error) {
+	var bytes []byte
+	var err error
+
+	fname := c.getResultCacheFilePath(netName, rt)
+	bytes, err = ioutil.ReadFile(fname)
+	if err != nil {
+		// Ignore read errors; the cached result may not exist on-disk
+		return nil, nil, nil
+	}
+
+	unmarshaled := cachedInfo{}
+	if err := json.Unmarshal(bytes, &unmarshaled); err != nil {
+		return nil, nil, fmt.Errorf("failed to unmarshal cached network %q config: %v", netName, err)
+	}
+	if unmarshaled.Kind != "cniCacheV1" {
```

Comment:
Mentioned above const should be also used there (while in tests it would be nice to still have a string value).